opengemini/point.go

```diff
@@ -97,69 +79,196 @@ func (bp *BatchPoints) AddPoint(p *Point) {
 	bp.Points = append(bp.Points, p)
 }
 
-func parseTags(tags map[string]string) string {
-	var builder strings.Builder
+type LineProtocolEncoder struct {
+	w io.Writer
+}
+
+func NewLineProtocolEncoder(w io.Writer) *LineProtocolEncoder {
+	return &LineProtocolEncoder{w: w}
+}
 
-	first := true
-	for k, v := range tags {
-		if !first {
-			builder.WriteByte(',')
-		} else {
-			first = false
+func (enc *LineProtocolEncoder) writeByte(b byte) error {
+	_, err := enc.w.Write([]byte{b})
+	return err
+}
+
+// writeString writes a string to the underlying writer, escaping characters
+// with a backslash if necessary. Note that, for simplity, `charsToEscape` can
+// only contain ASCII characters.
+func (enc *LineProtocolEncoder) writeString(s string, charsToEscape string) error {
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+
+		needEscape := strings.IndexByte(charsToEscape, c) != -1
+		if !needEscape && c == '\\' && i < len(s)-1 {
+			c1 := s[i+1]
+			needEscape = c1 == '\\' || strings.IndexByte(charsToEscape, c1) != -1
+		}
+
+		if needEscape {
+			if err := enc.writeByte('\\'); err != nil {
+				return err
+			}
+		}
+
+		if err := enc.writeByte(c); err != nil {
+			return err
 		}
-		builder.WriteString(k)
-		builder.WriteByte('=')
-		builder.WriteString(v)
 	}
 
-	return builder.String()
+	return nil
 }
 
-func parseFields(fields map[string]interface{}) string {
-	var builder strings.Builder
+func (enc *LineProtocolEncoder) writeFieldValue(v interface{}) error {
+	var err error
 
-	first := true
-	for k, v := range fields {
-		if !first {
-			builder.WriteByte(',')
-		} else {
-			first = false
-		}
-		builder.WriteString(k)
-		builder.WriteByte('=')
-		switch v := v.(type) {
-		case string:
-			builder.WriteString("\"" + v + "\"")
-		case int8, uint8, int16, uint16, int, uint, int32, uint32, int64, uint64:
-			builder.WriteString(fmt.Sprintf("%di", v))
-		case float32, float64:
-			builder.WriteString(fmt.Sprintf("%f", v))
-		case bool:
-			if v {
-				builder.WriteByte('T')
-			} else {
-				builder.WriteByte('F')
+	switch v := v.(type) {
+	case string:
+		if err = enc.writeByte('"'); err == nil {
+			if err = enc.writeString(v, `"`); err == nil {
+				err = enc.writeByte('"')
 			}
 		}
+	case int8:
+		if _, err = io.WriteString(enc.w, strconv.FormatInt(int64(v), 10)); err == nil {
```

Comment:
what do you think merge the same kind cases into one

Reply:
it is impossible unless we roll back to use `fmt.Printf` because `v` becomes an `interface{}` which `strconv.FormatXXX` does not accept.